Add Quirks for UA on iPad for tiktok.com

karlcow opened the pull request:
#### 6ee99ddf116e5a597fc14357479c523aee5ff29d
<pre>
Add Quirks for UA on iPad for tiktok.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=290372">https://bugs.webkit.org/show_bug.cgi?id=290372</a>
<a href="https://rdar.apple.com/146977044">rdar://146977044</a>

Reviewed by Brent Fulgham.

Currently TikTok is sniffing for Chrome UA in the
upload video section.
var e = navigator.userAgent.toLowerCase().match(/chrome\/(\d+)./);
And use this to display a message but without making
it explicit why this is needed. When adding a Chrome
string with a version number, the message can be
by-passed and the login is still working. This only
address the iPad side, a similar UA override needs
to be done for Safari on macOS.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsCustomUserAgentOverride):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies):

Canonical link: <a href="https://commits.webkit.org/293101@main">https://commits.webkit.org/293101@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7d5f959076cce346152d5b98e238b153434c2db

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97908 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17533 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7754 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103022 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48438 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99953 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17827 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25986 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74548 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31730 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100911 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13502 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88481 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54905 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13277 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6413 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47880 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83259 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6486 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105400 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24988 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83544 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25360 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84662 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82981 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20937 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27625 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5307 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18611 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24949 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24771 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28085 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26345 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1556,6 +1556,9 @@ static RecommendDesktopClassBrowsingForRequest desktopClassBrowsingRecommendedFo
         policies.setCustomUserAgent(standardUserAgentWithApplicationName(applicationName, emptyString(), UserAgentType::Desktop));
+        // FIXME: needs to be more generic so we can pass any domain with any strings modifications
+        if (Quirks::needsCustomUserAgentOverride(request.url()))
+            policies.setCustomUserAgent(makeStringByReplacingAll(standardUserAgentWithApplicationName(m_applicationNameForUserAgent), "like Gecko"_s, "like Gecko, like Chrome/136."_s));
```
I think this undoes the user agent changes made on line 1558. You probably need something like:

```
if (Quirks::needsCustomUserAgentOverride(request.url())) {
    auto customUserAgentQuirk = policies.customUserAgentAsSiteSpecificQuirks();
    if (!customUserAgentQuirk)
        customUserAgentQuirk = standardUserAgentWithApplicationName(m_applicationNameForUserAgent);
    policies.setCustomUserAgent(makeStringByReplacingAll(customUserAgentQuirk, "like Gecko"_s, "like Gecko, like Chrome/136."_s));
}
```

brentfulgham replied:
Karl and I spoke in person, and my code above is not right. Karl will update the patch with the right change. Please ignore the above.

karlcow replied:
@brentfulgham fixed.

brentfulgham reviewed (approved):
This looks fine, but we might want to try to capture `TikTok.com` in the global set of quirk flags so we don't have to do this string comparison every load. But this is probably a low-impact code path.

r=me if you fix the bug I showed in my comments below.

karlcow committed:
Add Quirks for UA on iPad for tiktok.com
https://bugs.webkit.org/show_bug.cgi?id=290372
rdar://146977044

Reviewed by Brent Fulgham.

Currently TikTok is sniffing for Chrome UA in the
upload video section.
var e = navigator.userAgent.toLowerCase().match(/chrome\/(\d+)./);
And use this to display a message but without making
it explicit why this is needed. When adding a Chrome
string with a version number, the message can be
by-passed and the login is still working. This only
address the iPad side, a similar UA override needs
to be done for Safari on macOS.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsCustomUserAgentOverride):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies):

Canonical link: https://commits.webkit.org/293101@main

brentfulgham commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1555,7 +1555,13 @@ static RecommendDesktopClassBrowsingForRequest desktopClassBrowsingRecommendedFo
+        // FIXME: This is done here for adding a UA override to tiktok.
+        // needsCustomUserAgentOverride() is currently very generic on purpose.
+        // In the future we want to pass more parameters for targeting specific domains.
+        if (Quirks::needsCustomUserAgentOverride(request.url()))
```
Oh no! We need to do the Quirks check again here. How do we disable this quirk otherwise?

karlcow replied:
rha. sorry about that. need to fix that. I will follow up with  https://bugs.webkit.org/show_bug.cgi?id=290953
